rclcpp_action/include/rclcpp_action/server.hpp

```diff
@@ -31,65 +36,428 @@ namespace rclcpp_action
+  /// \internal
+  RCLCPP_ACTION_PUBLIC
+  virtual
+  std::shared_ptr<void>
+  create_result_request() = 0;
+
+  /// Create an empty goal result message so it can be sent as a reply in a lower layer
+  /// \internal
+  RCLCPP_ACTION_PUBLIC
+  virtual
+  std::shared_ptr<void>
+  create_result_response(decltype(action_msgs::msg::GoalStatus::status) status) = 0;
+
+  /// \internal
+  RCLCPP_ACTION_PUBLIC
+  void
+  publish_status();
+
+  /// \internal
+  RCLCPP_ACTION_PUBLIC
+  void
+  publish_result(const std::array<uint8_t, 16> & uuid, std::shared_ptr<void> result_msg);
+
+  // End API for communication between ServerBase and Server<>
+  // ---------------------------------------------------------
 
 private:
+  /// Handle a request to add a new goal to the server
+  /// \internal
+  RCLCPP_ACTION_PUBLIC
+  void
+  execute_goal_request_received();
+
+  /// Handle a request to cancel goals on the server
+  /// \internal
+  RCLCPP_ACTION_PUBLIC
+  void
+  execute_cancel_request_received();
+
+  /// Handle a request to get the result of an action
+  /// \internal
+  RCLCPP_ACTION_PUBLIC
+  void
+  execute_result_request_received();
+
+  /// Handle a timeout indicating a completed goal should be forgotten by the server
+  /// \internal
+  RCLCPP_ACTION_PUBLIC
+  void
+  execute_check_expired_goals();
+
+  /// Private implementation
+  /// \internal
   std::unique_ptr<ServerBaseImpl> pimpl_;
 };
 
 
-/// Templated Action Server class
-/// It is responsible for getting the C action type support struct from the C++ type, and
-/// calling user callbacks with goal handles of the appropriate type.
+/// Hash a goal id so it can be used as a key in std::unordered_map
+struct UUIDHash
+{
+  size_t operator()(std::array<uint8_t, 16> const & uuid) const noexcept
+  {
+    // TODO(sloretz) Use someone else's hash function and cite it
+    size_t result = 0;
+    for (size_t i = 0; i < 16; ++i) {
+      for (size_t b = 0; b < sizeof(size_t); ++b) {
+        size_t part = uuid[i];
+        part <<= CHAR_BIT * b;
+        result ^= part;
+      }
+    }
+    return result;
+  }
+};
+
+/// Action Server
+/**
+ * This class creates an action server.
+ *
+ * Create an instance of this server using `rclcpp_action::create_server()`.
+ *
+ * Internally, this class is responsible for:
+ *  - coverting between the C++ action type and generic types for `rclcpp_action::ServerBase`, and
+ *  - calling user callbacks.
+ */
 template<typename ACTION>
 class Server : public ServerBase
 {
 public:
   RCLCPP_SMART_PTR_DEFINITIONS_NOT_COPYABLE(Server)
 
-  using GoalCallback = std::function<void (rcl_action_goal_info_t &, typename ACTION::Goal *)>;
-  using CancelCallback = std::function<void (std::shared_ptr<ServerGoalHandle<ACTION>>)>;
+  /// Signature of a callback that accepts or rejects goal requests.
+  using GoalCallback = std::function<GoalResponse(
+        std::array<uint8_t, 16>&, std::shared_ptr<typename ACTION::Goal>)>;
```

Comment:
Why is this shared pointer not `const` here, but it is for `get_goal()` on the `ServerGoalHandle<ActionT>`? Should it be const here too, or is this a copy of the goal?

Reply:
I think it should be `const`

Reply:
const in b1828727242a48d6206b1b97307c5bf49d626a89